common/lib/xmodule/xmodule/modulestore/tests/test_publish.py

```diff
@@ -139,3 +152,614 @@ def test_publish_draft_delete(self):
+            elif block_type == 'html':
+                regex_func = self._make_html_unit_matching_regex
+            block_regex = regex_func(
+                self.course_key,
+                draft=draft,
+                **block_params
+            )
+            self.assertOLXContent(block_type, block_id, draft=draft, xml_re=block_regex)
+            self.assertOLXMissing(block_type, block_id, draft=(not draft))
+
+    def assertOLXIsDraft(self, block_list, **kwargs):
+        """
+        Check that all blocks in the list are draft blocks in the OLX format when the course is exported.
+        """
+        self._assertOLXBase(block_list, draft=True, **kwargs)
+
+    def assertOLXIsPublished(self, block_list, **kwargs):
+        """
+        Check that all blocks in the list are published blocks in the OLX format when the course is exported.
+        """
+        self._assertOLXBase(block_list, draft=False, **kwargs)
+
+    def publish(self, block_type, block_id):
+        """
+        Get an item, publish it, and shift to a new course export dir.
+        """
+        # Get the specified test item from the draft branch.
+        with self.store.branch_setting(ModuleStoreEnum.Branch.draft_preferred):
+            test_item = self.store.get_item(self.course_key.make_usage_key(block_type=block_type, block_id=block_id))
+        # Publish the draft item to the published branch.
+        self.store.publish(test_item.location, self.user_id)
+        # Since the elemental operation is now complete, shift to the post-operation export directory name.
+        self.export_dir = self.EXPORTED_COURSE_AFTER_DIR_NAME
+        self.before_publish = False
+
+
+@ddt.ddt
+class ElementalPublishingTests(UniversalTestProcedure):
+    """
+    Tests for the publish() operation.
+    """
+    @ddt.data(DRAFT_MODULESTORE_SETUP, MongoModulestoreBuilder())
+    def test_publish_old_mongo_unit(self, modulestore_builder):
+        with self._setup_test(modulestore_builder, self.course_key) as (
+            self.root_export_dir, self.contentstore, self.store, self.course
+        ):
+            block_type = 'html'
+            block_id = 'unit00'
+            # In old Mongo, you can successfully publish an item whose parent
+            # isn't published.
+            self.publish(block_type, block_id)
+
+    @ddt.data(SPLIT_MODULESTORE_SETUP)
+    def test_publish_split_unit(self, modulestore_builder):
+        with self._setup_test(modulestore_builder, self.course_key) as (
+            self.root_export_dir, self.contentstore, self.store, self.course
+        ):
+            block_type = 'html'
+            block_id = 'unit00'
+            # In Split, you cannot publish an item whose parents are unpublished.
+            # Split will raise an exception when the item's parent(s) aren't found
+            # in the published branch.
+            with self.assertRaises(ItemNotFoundError):
+                self.publish(block_type, block_id)
```

Comment:
So, this is one place where the semantics of the API differs. Not for this PR, but it would be good to get to the point where the semantics are consistent.